### ISSUE TITLE
Temperature Over Time

### ISSUE THREAD

**Issue by Freyert** [ClementTsang/bottom#1366]
### Checklist

- [X] I've looked through [the documentation](https://clementtsang.github.io/bottom/nightly/) and  [existing open issues](https://github.com/ClementTsang/bottom/issues?q=is%3Aopen+is%3Aissue+label%3Afeature) for similar feature requests.

### Describe the feature request

My motivation is that I have issues on my computer where it overheats, and crashes. I'm actually not sure it's overheating though.

I was looking at the code and readme which caused me to notice that there are essentially Widgets with Graphs and Widgets without Graphs.

I also noticed you doing a larger rewrite of the widget system to improve extensibility https://github.com/ClementTsang/bottom/issues/374

I think the core issue why I can't graph temperatures is related to this lack of extensibility.

From my personal exploration:
* The application collects data into  https://github.com/ClementTsang/bottom/blob/main/src/data_collection.rs#L29 for rendering.
* There is a specific widget for CPU https://github.com/ClementTsang/bottom/blob/main/src/widgets/cpu_graph.rs
* There is a specific widget for Temperature https://github.com/ClementTsang/bottom/blob/main/src/widgets/temperature_table.rs

My expectations would be that I could convert either one of these metrics into an instantaneous table or a graph as suits me. Though I can also understand particular widgets like the process table being different...

**Comment by ClementTsang** [comment]
Yeah, something I wanted to do (I made a comment a while ago in https://github.com/ClementTsang/bottom/issues/205), but haven't had time with other stuff in life.

**Comment by Freyert** [comment]
@ClementTsang I see! That's helpful to get more context from those other issues.

I think I'll fork and make a mess, but try some things out. If I come up with anything I'll share it back here for your consideration.

**Comment by Freyert** [comment]
Ok so my main focus area is the fact that all of these widgets for graphs have to implement a different draw function depending on the data they receive.

```mermaid
flowchart TD
    subgraph src/canvas.rs
    A[struct Painter] --> B[impl Painter.draw_data]
    end

    subgraph src/canvas/widgets/cpu_graph.rs
    B --> C[impl Painter.draw_cpu]
    end

    subgraph src/canvas/widgets/mem_graph.rs
    B --> D[impl Painter.draw_memory_graph]
    end

    subgraph src/canvas/widgets/network_graph.rs
    B --> E[impl Painter.draw_network]
    end
```

**Comment by Freyert** [comment]
My plan is to invert control here a bit. Instead of a CPU Widget or a Network Widget... We can have a Graph Widget or a Table Widget.

So then where do we get our data from? From things that are Graphable or Tableable. Which now I think is trapped in the widget state.

**Comment by Freyert** [comment]
I've been exploring more and it seems like separation of concerns is a bit difficult. There are good reasons for everything to be scoped the way it is, but there are ways to separate concerns with a little more added complexity.

---

One of my big goals is to have the `state`  objects (`TempState`, `NetState`) _own_ the data that their widgets need to render.

If we're talking about a Model View Controller framework (which bottom uses?) then the `state` objects should own all data that the application uses.

---

However, we also have the concept of the `Collector`. The `Collector` runs in a separate thread to collect system metrics on a defined interval.

I first thought it would be good to separate collection to each `Widget` so that each `Widget` would have a `State` and a `Collector`. However this isn't ideal because `Collector` collects `sysinfo` for example and then delivers it to several other widgets.

It would be inefficient at this moment to have each widget have a thread for collecting their info because they would all make redundant calls with `sysinfo`.

---

This pattern could be handled by [Channels](https://doc.rust-lang.org/rust-by-example/std_misc/channels.html). However we would now need a thread to pull data into the state objects.

`collect -> channel_n -> Vec<i32>`.

Downside: need thread to pull data out of the channel.
Upside: it probably already exists.

**Comment by NiciTheNici** [comment]
heya, I'm interested in this too. Got some experience with rust and might be able to hack something together if I find the time (you know how it is...)

I'm thinking we could reuse the cpu view and graph...
- ...toggling between usage and the temps via some shortcut
- ...as well as a flag to start bottom in temps mode
- ...showing the cpu usage in the current temps widget when toggled might also be neat.

**Comment by ClementTsang** [comment]
Hm, no, I would not want to do that. I would prefer to add a separate widget in general.

It's not too hard, but I just haven't had time to do this, sorry! It's on my todo for the near term.

**Comment by elazarl** [comment]
I guess the same is relevant for the disk widget, as rearchitecturing the widget would solve both, right?

**Comment by elazarl** [comment]
I see [that there's a placeholder for that](https://github.com/ClementTsang/bottom/pull/1733/files#diff-cb762aed9c1c4f4210fc825f9640640f5551791fb5fd66e34626bfe942c35459R1)
